src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -775,18 +763,57 @@ private static void finish(@CheckForNull final String cookie) {
         private static final class Callback extends BodyExecutionCallback.TailCall {
 
             private final String cookie;
+            @Deprecated
             private WorkspaceList.Lease lease;
+            private final ExecutorStepExecution execution;
 
-            Callback(String cookie, WorkspaceList.Lease lease) {
+            Callback(String cookie, ExecutorStepExecution execution) {
                 this.cookie = cookie;
-                this.lease = lease;
+                this.execution = execution;
             }
 
             @Override protected void finished(StepContext context) throws Exception {
                 LOGGER.log(FINE, "finished {0}", cookie);
-                lease.release();
-                lease = null;
-                finish(cookie);
+                try {
+                    if (execution != null) {
+                        WorkspaceList.Lease _lease = ExtensionList.lookupSingleton(ExecutorStepDynamicContext.WorkspaceListLeaseTranslator.class).get(execution.state);
+                        if (_lease != null) {
+                            _lease.release();
+                        }
+                    } else {
+                        lease.release();
+                        lease = null;
+                    }
+                } finally {
+                    finish(cookie);
+                }
+                if (execution != null) {
+                    boolean _stopping = execution.state.task.stopping;
+                    execution.state.task.stopping = true;
+                    try {
+                        Queue.getInstance().cancel(execution.state.task);
+                    } finally {
+                        execution.state.task.stopping = _stopping;
+                    }
+                    execution.state = null;
+                }
+            }
+
+            @Override public void onFailure(StepContext context, Throwable t) {
+                try {
+                    if (execution != null) {
+                        TaskListener listener = context.get(TaskListener.class);
+                        if (ExtensionList.lookup(ExecutorStepRetryEligibility.class).stream().anyMatch(e -> e.shouldRetry(t, execution.state.node, execution.step.getLabel(), listener))) {
+                            finished(context);
+                            execution.start();
+                            return;
+                        }
+                        listener.getLogger().println("No plugin requested a retry of a failed node block running on " + execution.state.node);
```

Comment:
nit: Seems kind of unusual to me to put this message in the build log for users, since as far as I understand it is the current and expected behavior.

Reply:
Perhaps so.

Reply:
But in practice, if retry is normally accepted but in some case is rejected, you want to know why. In fact it might be necessary to print the exception message or something. Needs refinement.